File: [B]linked-list/[B]linked-list/234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return head

        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        first_half_tail = slow
        second_half_head = first_half_tail.next

        prev = None
        cur = second_half_head
        while cur:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        new_second_half_head = prev

        p1 = head
        p2 = new_second_half_head
        while p2:
            if p1.val != p2.val:
                return False
            p1 = p1.next
            p2 = p2.next
        return True
        
# time O(n)
# space O(1)
# using linked lsit and two pointers (utilize symmetry property) (finding middle, reversing, comparing)

File: [B]linked-list/[B]linked-list/234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode firstHalfTail = slow;
        ListNode secondHalfHead = firstHalfTail.next;

        ListNode prev = null;
        ListNode cur = secondHalfHead;
        while (cur != null) {
            ListNode nxt = cur.next;
            cur.next = prev;
            prev = cur;
            cur = nxt;
        }
        ListNode newSecondHalfHead = prev;
        ListNode p1 = head;
        ListNode p2 = newSecondHalfHead;
        while (p2 != null) {
            if (p1.val != p2.val) {
                return false;
            }
            p1 = p1.next;
            p2 = p2.next;
        }
        return true;
    }
}

// time O(n)
// space O(1)
// using linked lsit and two pointers (utilize symmetry property) (finding middle, reversing, comparing)